### PR TITLE
add assume unchanged support. It also improves the skip worktree.

### PR DESCRIPTION
**Pull request by zilongshanren**
resolve issue https://github.com/magit/magit/issues/531

I have added two keybindings, it might not be optimal, helps arewanted.

thanks.

### CODE REVIEW

File: lisp/magit-git.el [hunk 1]
```diff
@@ -828,6 +828,13 @@ tracked file."
 (defun magit-file-tracked-p (file)
   (magit-git-success "ls-files" "--error-unmatch" file))
 
+ 
+(defun magit-list-all-files ()
```
Comment (tarsius) [inline lisp/magit-git.el]: How is this different from `magit-list-files` except for the use of `magit-with-toplevel` and lacking the ARGS argument?

IMO it is the callers that should use `magit-with-toplevel` if appropriate. The difference in the names of these functions, i.e. whether `all-` appears in it or not, doesn't hint at what the differences between the functions are.

Is the `magit-with-toplevel` really necessary in

File: lisp/magit-git.el [hunk 2]
```diff
@@ -861,10 +868,21 @@ tracked file."
...
+                           "-v" "--full-name" "-z")))
+
 (defun magit-skip-worktree-files ()
   (--keep (and (and (= (aref it 0) ?S)
                     (substring it 2)))
-          (magit-git-items "ls-files" "-z" "-t")))
+          (magit-git-items "ls-files"
+                           (string-trim
+                            (car (magit-git-items "rev-parse" "--show-toplevel")))
+                           "-z" "--full-name" "-t")))
```
Comment (tarsius) [inline lisp/magit-git.el]: Do this in a separate commit and explain what it does.

File: lisp/magit-gitignore.el [hunk 3]
```diff
@@ -130,22 +130,42 @@ Rules that are defined in that file affect all local repositories."
...
+                                              (magit-list-all-files)
                                               (magit-skip-worktree-files)))))
   (magit-with-toplevel
     (magit-run-git "update-index" "--skip-worktree" "--" file)))
 
 ;;;###autoload
 (defun magit-no-skip-worktree (file)
   "Call \"git update-index --no-skip-worktree FILE\"."
-  (interactive (list (magit-read-file-choice "Do not skip worktree for: "
+  (interactive (list (magit-read-file-choice "Do not skip worktree for"
```
Comment (tarsius) [inline lisp/magit-gitignore.el]: Do this and the other removal of `: ` in a separate commit.

File: lisp/magit-status.el [hunk 4]
```diff
@@ -685,6 +687,22 @@ of that variable can be set using \"D -- DIRECTORY RET g\"."
         (magit-insert-files files base)
         (insert ?\n)))))
 
+(defun magit-insert-assume-unchanged-files ()
+  "Insert a tree of assume unchanged files.
+
+    If the first element of `magit-buffer-diff-files' is a
+    directory, then limit the list to files below that.  The value
+    of that variable can be set using \"D -- DIRECTORY RET g\"."
```
Comment (tarsius) [inline lisp/magit-status.el]: Do not indent doc-strings. (Look at this using `describe-variable` to see why.)

File: lisp/magit-git.el (hunk 1 above)
Comment (zilongshanren) [reply lisp/magit-git.el]: `magit-list-all-files` will list all the files in the git repository regardless where you run the function , while the `magit-list-files` only list files under current buffer directory.

With the `magit-with-toplevel` wrapper, the git commands could be executed in the root of git repository.

File: lisp/magit-git.el (hunk 2 above)
Comment (zilongshanren) [reply lisp/magit-git.el]: @tarsius Sure. I could split the commit into two speparate commits.

It means run `git ls-files` in the root dir of a git repository.

File: lisp/magit-status.el (hunk 4 above)
Comment (zilongshanren) [reply lisp/magit-status.el]: @tarsius Ok, I don't notice the difference here and just bindly re-indent the doc string here.
I will take a look at what it behaviors using `describe-variable`

**Comment by tarsius** [comment]
The obvious part of this is to copy the relevant skip-worktree code and doing `s/skip-worktree/assume-unchanged`.  There are some additional details you overlooked though. You also noticed that there was an issue with paths, regarding sub-directories.  I do not agree with the way you addressed those issue.

I decided it would be easier to make the changes myself instead of guiding you to do so.  I did however try to do it in a away that allows you to see what different decisions I came to and why.

I did not add the section inserters to the the hook (you added it to the wrong hook in your commit).

**Comment by zilongshanren** [comment]
@tarsius Alright.

**Comment by phil-s** [comment]
Thanks for the work on this handy new feature @zilongshanren.

@tarsius, I didn't see key bindings for the new commands in the commits, but I feel that they would logically fit under the "ignore" menu?

**Comment by tarsius** [comment]
> logically fit under the "ignore" menu?

Done.

**Comment by zilongshanren** [comment]
should we mention adding the ignored section in magit status buffer? I think it's more obvious when people mark some files as unchanged/skip worktree and they can see these files in the status buffer.

```      
  (magit-add-section-hook 'magit-status-sections-hook
                                'magit-insert-assume-unchanged-files nil t)

        ;; insert the hidden files section in the magit status buffer.
        (magit-add-section-hook 'magit-status-sections-hook
                                'magit-insert-skip-worktree-files nil t)
```

**Comment by tarsius** [comment]
Done that too.
